rclcpp/include/rclcpp/service.hpp

```diff
@@ -118,16 +116,25 @@ class Service : public ServiceBase
     using rosidl_typesupport_cpp::get_service_type_support_handle;
     auto service_type_support_handle = get_service_type_support_handle<ServiceT>();
 
+    std::weak_ptr<rcl_node_t> weak_node_handle(node_handle_);
     // rcl does the static memory allocation here
     service_handle_ = std::shared_ptr<rcl_service_t>(
       new rcl_service_t, [ = ](rcl_service_t * service)
       {
-        if (rcl_service_fini(service, node_handle_.get()) != RCL_RET_OK) {
-          RCLCPP_ERROR(
-            rclcpp::get_logger(rcl_node_get_name(node_handle.get())).get_child("rclcpp"),
-            "Error in destruction of rcl service handle: %s",
-            rcl_get_error_string_safe());
-          rcl_reset_error();
+        auto handle = weak_node_handle.lock();
+        if (handle) {
+          if (rcl_service_fini(service, handle.get()) != RCL_RET_OK) {
+            RCLCPP_ERROR(
+              rclcpp::get_logger(rcl_node_get_name(handle.get())).get_child("rclcpp"),
+              "Error in destruction of rcl service handle: %s",
+              rcl_get_error_string_safe());
+            rcl_reset_error();
+          }
+        } else {
+          RCLCPP_FATAL(
```

Comment:
changed these to `RCLCPP_FATAL` (there is no RCLCPP_CRITICAL`). Let me know if you would rather have it classified differently.

Reply:
Sure thing.